guardian/elastic-search-monitor
# Migrate CloudWatch alarms into cdk

## What does this change?

In https://github.com/guardian/elastic-search-monitor/pull/17, we migrated all of the application infrastructure to `@guardian/cdk`, but we [left some alarm configuration in standard CloudFormation](https://github.com/guardian/elastic-search-monitor/pull/17#discussion_r847122370).

This PR ports the alarms into TypeScript/`cdk` and removes the remaining CloudFormation from this repository.

I decided to move the alarms into the same stack as the application infrastructure (previously we had a separate stack for alarms), which means that they are now deployable via Riff-Raff (we gain CD for alarms as part of this PR). Once this PR has been merged we can delete the `elastic-search-monitor-alarms-PROD` stack (which is based on https://github.com/guardian/elastic-search-monitor/pull/19/commits/a2daa755647c49ef48675d8d85f33803ab4a834e). 

## How to test

I've deployed this to `PROD` (there is no `CODE` environment for this app) to confirm that this change can be deployed as expected. 

I also used [`describe-alarms`](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/cloudwatch/describe-alarms.html) and `diff -y` to compare the old alarms with the new alarms.

I can see that the only fields which change are expected (`AlarmName`, `AlarmArn`, `AlarmConfigurationUpdatedTimestamp`, `StateReason`, `StateReasonData`, `StateUpdatedTimestamp`). 

This technique also helped me to catch some drift on the original stack (see https://github.com/guardian/elastic-search-monitor/pull/19/commits/d82e92ec011977e8ee150b321e7e48c895a51a4a) - these (helpful!) manual updates would have been lost if we'd just compared the old CloudFormation with the generated CloudFormation.

## How can we measure success?

* Less cognitive overhead (all of the infrastructure is defined in `cdk`, rather having than a mix of `cdk` and CloudFormation)
* It's _much_ easier to work with alarms in TypeScript (in my opinion)
* Changes to the alarms can now be deployed by Riff-Raff, just like all other infrastructure

## Have we considered potential risks?

Yes, since we create brand new alarms in this PR, there is a risk that they won't work properly; I've tried to mitigate that risk by comparing the alarms carefully!
